### PR TITLE
http3: process 1xx status codes

### LINKED ISSUE #4403: http3.RoundTripper unable to process 1xx response
When website enables [HTTP 103 Early Hints](https://http.dev/103), `http3.RoundTripper` returned "wrong?" status code and client failed to read response body

This code can reproduce this: 
 - http2 client returns status 307 and correct body (and `httptrace Got1xxResponse` was called before roundtrip returned)
 - http3 client returns status 103 and empty body

```go
package xxx

import (
	"context"
	"net/http"
	"net/http/httptrace"
	"net/textproto"
	"testing"

	"github.com/quic-go/quic-go/http3"
	"golang.org/x/net/http2"
)

func TestHttp3(t *testing.T) {
	ctx := context.Background()
	req, _ := http.NewRequestWithContext(
		httptrace.WithClientTrace(ctx, &httptrace.ClientTrace{
			Got1xxResponse: func(code int, header textproto.MIMEHeader) error {
                                 // do something
				t.Log("got 1xx response", code, header) // <-- ✅ http/2 output status 103
				return nil
			},
		}),
		http.MethodGet,
		"https://betterprogramming.pub/object-oriented-programming-the-trillion-dollar-disaster-92a4b666c7c7",
		nil,
	)
	req.Header.Set("User-Agent", "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/123.0.0.0 Safari/537.36")

	tr2 := &http2.Transport{}
	resp2, _ := tr2.RoundTrip(req)
	braw2, _ := io.ReadAll(resp2.Body)
	t.Log(resp2.StatusCode, string(braw2)) // <-- ✅ http/2 output status 307 and content `<a href="https: ......`

	rr3 := &http3.RoundTripper{}
	resp3, _ := rr3.RoundTrip(req)
	braw3, _ := io.ReadAll(resp3.Body)
	t.Log(resp3.StatusCode, string(braw3)) // <-- ❓http/3 output status 103 and empty content
}
```

`curl -v https://xxx --http2` output:
```bash
> GET /..... HTTP/2
> Host: ......
>
......
< HTTP/2 103
< link: <https://......
<
< HTTP/2 307
< date: ......
< content-type: text/html; charset=utf-8
......
<
<a href="https://......
```

### PR DESCRIPTION
**Pull request by mchtech**
Fix #4403
copy the impletation from the go standard library and add tests.

### CODE REVIEW

File: http3/client.go [hunk 1]
```diff
@@ -300,9 +303,37 @@ func (c *SingleDestinationRoundTripper) doRequest(req *http.Request, str quic.St
...
-	res, err := hstr.ReadResponse()
-	if err != nil {
-		return nil, err
+	var (
+		res *http.Response
+		err error
+	)
+
+	// copy from net/http: support 1xx responses
+	trace := httptrace.ContextClientTrace(req.Context())
```
Comment (marten-seemann) [inline http3/client.go:312]: We don't support tracing yet.

**Review by marten-seemann — COMMENTED** [review]
Can you add an integration test as well? `integrationtests/self/http_test.go`

**Comment by marten-seemann** [comment]
@mchtech Are you planning to address the review comments here? Would be great if we could include this fix in the next release!

File: http3/client.go (hunk 1 above)
Comment (marten-seemann) [reply http3/client.go:312]: Actually, this is fine. Let's keep it as is.

**Comment by mchtech** [comment]
I will add an integration test in this weekend.

**Review by marten-seemann — APPROVED** [review]
Thank you!

File: http3/client_test.go [hunk 2]
```diff
@@ -858,5 +864,75 @@ var _ = Describe("Client", func() {
...
+				rspBuf := bytes.NewBuffer(encodeResponse(103))
+				gomock.InOrder(
+					conn.EXPECT().HandshakeComplete().Return(handshakeChan),
+					conn.EXPECT().OpenStreamSync(ctx).Return(str, nil),
+					conn.EXPECT().ConnectionState().Return(quic.ConnectionState{}),
+				)
+				str.EXPECT().Write(gomock.Any()).AnyTimes().DoAndReturn(func(p []byte) (int, error) { return len(p), nil })
+				str.EXPECT().Close()
+				str.EXPECT().Read(gomock.Any()).DoAndReturn(rspBuf.Read).AnyTimes()
+				rsp, err := cl.RoundTripOpt(req, RoundTripOpt{})
```
Comment (marten-seemann) [inline http3/client_test.go:895]:
```suggestion
				rsp, err := cl.RoundTrip(req)
```

File: http3/client_test.go [hunk 3]
```diff
@@ -858,5 +864,75 @@ var _ = Describe("Client", func() {
...
+				rspBuf := bytes.NewBuffer(encodeResponse(101))
+				gomock.InOrder(
+					conn.EXPECT().HandshakeComplete().Return(handshakeChan),
+					conn.EXPECT().OpenStreamSync(ctx).Return(str, nil),
+					conn.EXPECT().ConnectionState().Return(quic.ConnectionState{}),
+				)
+				str.EXPECT().Write(gomock.Any()).AnyTimes().DoAndReturn(func(p []byte) (int, error) { return len(p), nil })
+				str.EXPECT().Close()
+				str.EXPECT().Read(gomock.Any()).DoAndReturn(rspBuf.Read).AnyTimes()
+				rsp, err := cl.RoundTripOpt(req, RoundTripOpt{})
```
Comment (marten-seemann) [inline http3/client_test.go:927]:
```suggestion
				rsp, err := cl.RoundTrip(req)
```

**Pull request merged**